---
title: "RNAseq"
author: "Lakshmi"
date: "2024-07-05"
output: html_document
---

# Basic tutorial on RNASeq

## Installing and Loading the packages
```{r, message = FALSE}
# All packages need to be downloaded

# install.packages(tidyverse)
# install.packages(ggplot)
# install.packages(ggrepel)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("DESeq2")
# BiocManager::install("ComplexHeatmap")
# BiocManager::install("clusterProfiler")
# BiocManager::install("org.Hs.eg.db")

## Loading all the necessary packages
library(tidyverse) #data
library(ggplot2) #plot
library(ggrepel) #plot

library(DESeq2) #ngs
library(ComplexHeatmap) #plot
library(clusterProfiler) #pathway
library(org.Hs.eg.db) #pathway
```

## load data 
Its a count matrix and meta data. The process of how rna reads became count matrix is a black box for now. Will work on it after this. 
```{r}
counts <- read.csv("./data/tutorial_counts.csv", check.names = FALSE)
metadata <- read.csv("./data/tutorial_metadata.csv", row.names = 1)
```

This basically sets the basis of the analysis, with "Mock" being level 1 and "Virus being level 2
```{r}
str(metadata)
metadata$Strain = factor(metadata$Strain)
metadata$Strain = relevel(metadata$Strain, ref = "Mock")
```

This analysis below shows that the count matrix has duplicated "gene names" but in actual it is all "-" genes without names
```{r}
head(counts)
dim(counts)
length(unique(counts$GeneSymbol))
counts[duplicated(counts$GeneSymbol),]
```

Here, we have piped counts to do a grouping by gene symbol, then summarizing "across" every column ("everything") using mean. Now, its no more duplicated. However, note that this made the "counts" float. Need to convert this to integer. 
```{r}
avg_counts = counts %>% group_by(GeneSymbol) %>% summarise(across(everything(),mean))
sum(duplicated(avg_counts$GeneSymbol))
avg_counts = avg_counts %>% mutate(across(where(is.numeric),as.integer)) #cant do as.integer(avg_counts) as its a tibble not a base data.frame. This also lets go of the gene symbol column (str), and now we have only 6 columns. 
```

## DESeq2

Now, we are ready for DESeq Analysis 
```{r}
dds = DESeqDataSetFromMatrix(avg_counts, colData = metadata, design = ~ Strain) #relationship between all the variables (~) with Strain 

```






